src/validation/rules/package_name.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 Helsing GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+/// Ensure that the protobuf package names match the buffrs package name.
+#[derive(Debug, Clone)]
+pub struct PackageName {
+    /// Package name to enforce.
+    name: String,
+}
+
+impl PackageName {
+    /// Create new checker for this rule.
+    pub fn new(name: &str) -> Self {
+        Self { name: name.into() }
+    }
+}
+
+fn is_prefix(prefix: &str, package: &str) -> bool {
+    prefix
+        .split('.')
+        .zip(package.split('.'))
+        .all(|(a, b)| a == b)
+}
+
+impl Rule for PackageName {
+    fn rule_info(&self) -> &'static str {
+        "Make sure that the protobuf package name matches the buffer package name."
+    }
+
+    fn check_package(&mut self, package: &Package) -> Violations {
```

Comment:
It worries me that this takes a mutable reference to self though

Reply:
That is intentional so that rules can have internal state. We can address this in a future PR.